### ISSUE TITLE
Placeholder parameters should be allowed in trait fns

### ISSUE THREAD

**Issue by regiontog** [rust-analyzer/rust-analyzer#597]
The [following](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a1915259c1d4ece0c58eb5d6c9b58a85) (valid) code results in errors in the [syntax tree](https://gist.github.com/regiontog/f83044bd6ef2018150fa6f4aa6de9d3e#file-gistfile1-txt-L20).

```rust
trait Foo {
    fn bar(_: bool);
}
```

**Comment by regiontog** [comment]
During parsing of fn defs inside traits `value_parameter` gets called with `flavor` `Flavor::OptionalType` as far as I can tell.

https://github.com/rust-analyzer/rust-analyzer/blob/595a2f9900d052d0c5cc3529602aec0cbcd52614/crates/ra_syntax/src/grammar/params.rs#L66-L94

I don't understand why a pattern is parsed here, should not this be an identifier?

**Comment by matklad** [comment]
>I don't understand why a pattern is parsed here, should not this be an identifier?

No, it can be a full pattern in general: `fn foo((x, y): (i32, u64))`.

However in trait methods we used to (pre-2018) allow omitting the name of the type, and so (to keep parsing LL(k)), only a small non-recursive set of patterns is allowed in that case. `OptionalPattern` is exactly the branch which deals with this curiosity.

**Comment by matklad** [comment]
Specifically, `fn foo(&x: &i32)` is valid as a trait method, and `fn foo(&&x: &&i32)` is, currently, not valid.

**Comment by matklad** [comment]
So, we need to change `la0 == IDENT` to `(la0 == IDENT || la0 == UNDERSCORE)` I think?

**Comment by regiontog** [comment]
> 
> 
> So, we need to change `la0 == IDENT` to `(la0 == IDENT || la0 == UNDERSCORE)` I think?

Yep, that fixes it as far as I can see 😃
